### PR TITLE
Promote stack management to the JsonReader/JsonWriter supertypes.

### PR DESCRIPTION
**Pull request by swankjesse**
It turns out that we can reuse a lot of code with inheritance. Who knew?

### CODE REVIEW

File: moshi/src/test/java/com/squareup/moshi/JsonWriterPathTest.java [hunk 1]
```diff
@@ -17,14 +17,27 @@
...
+import org.junit.runners.Parameterized;
+import org.junit.runners.Parameterized.Parameter;
+import org.junit.runners.Parameterized.Parameters;
 
 import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.Assume.assumeTrue;
 
+@RunWith(Parameterized.class)
 public final class JsonWriterPathTest {
+  @Parameter public JsonWriterFactory factory;
```
Comment (swankjesse) [inline moshi/src/test/java/com/squareup/moshi/JsonWriterPathTest.java]: now this test operates on both DOMs and bytes

File: moshi/src/main/java/com/squareup/moshi/JsonReader.java [hunk 2]
```diff
@@ -349,7 +394,9 @@ public static JsonReader of(BufferedSource source) {
    * Returns a <a href="http://goessner.net/articles/JsonPath/">JsonPath</a> to
    * the current location in the JSON value.
    */
-  public abstract String getPath();
+  public String getPath() {
```
Comment (JakeWharton) [inline moshi/src/main/java/com/squareup/moshi/JsonReader.java]: Should these all be `final` then?

File: moshi/src/main/java/com/squareup/moshi/JsonWriter.java [hunk 3]
```diff
@@ -276,5 +322,7 @@ public static JsonWriter of(BufferedSink sink) {
    * Returns a <a href="http://goessner.net/articles/JsonPath/">JsonPath</a> to
    * the current location in the JSON value.
    */
-  public abstract String getPath();
+  public String getPath() {
```
Comment (JakeWharton) [inline moshi/src/main/java/com/squareup/moshi/JsonWriter.java]: Same `final` question.

File: moshi/src/main/java/com/squareup/moshi/JsonReader.java (hunk 2 above)
Comment (swankjesse) [inline moshi/src/main/java/com/squareup/moshi/JsonReader.java]: Fixed ’em all.

**Pull request merged**